# Homework #8

**Due by midnight on Friday April 26, 2019**. Complete all of the following problems. Ideally, the theoretical problems should be answered in a Markdown cell directly underneath the question. If you don't know LaTex/Markdown, you may submit separate handwritten solutions to the theoretical problems. Please do not turn in messy work. Computational problems should be completed in this notebook (using the R kernel is preferred). Computational questions may require code, plots, analysis, interpretation, etc. Working in small groups is allowed, but it is important that you make an effort to master the material and hand in your own work. 


Name: Adam Ten Hoeve

## A. Theoretical Problems

## A.1 

#### (a) In the context of simple linear regression, show that if we center both $Y$ and $X$, then $\widehat\beta_0 = 0$. 

For SLR, we want to estimate $\hat{\alpha}$ and $\hat{\beta}$ such that:

$$
\begin{align*}
    0 &= \sum_{i=1}^n [y_i - \hat{\beta_1}*x_i - \hat{\beta_0}] \\
    0 &= -n\hat{\beta_0} + \sum_{i=1}^n [Y_i - \hat{\beta_1}*x_i] \\
    \hat{\beta_0} &= \frac{1}{n}\sum_{i=1}^n [y_i - \hat{\beta_1}*x_i] \\
\end{align*}
$$

If we center both Y and X, then we would be subtracting the mean from those vectors. Notice that:

$$
\begin{align*}
\sum_{i=1}^n(x_i - \bar{x}) &= \sum_{i=1}^n(x_i) - \sum_{i=1}^n(\bar{x}) \\
&= \sum_{i=1}^n(x_i) - \sum_{i=1}^n \Big( \frac{1}{n} \sum_{i=1}^{n}(x_i) \Big) \\
&= \sum_{i=1}^n(x_i) - \sum_{i=1}^n(x_i) \\
&= 0
\end{align*}
$$

This is also true for Y and $\bar{y}$. Going back to our first equation we get:

$$
\begin{align*}
\hat{\beta_0} &= \frac{1}{n}\sum_{i=1}^n (y_i - \bar{y}) - \hat{\beta_1}*(x_i - \bar{x}) \\
\hat{\beta_0} &= \frac{1}{n}\sum_{i=1}^n (y_i - \bar{y}) - \hat{\beta_1}*\frac{1}{n}\sum_{i=1}^n(x_i - \bar{x}) \\
\hat{\beta_0} &= 0 - 0
\end{align*}
$$

Therefor, the intercept of SLR is 0 when both X and Y are centered.

#### (b) Suppose that we only center $Y$ but not $X$. What does $\widehat\beta_0$ become?

$$
\begin{align*}
\hat{\beta_0} &= \frac{1}{n}\sum_{i=1}^n (y_i - \bar{y}) - \hat{\beta_1}*x_i \\
\hat{\beta_0} &= \frac{1}{n}\sum_{i=1}^n (y_i - \bar{y}) - \hat{\beta_1}*\frac{1}{n}\sum_{i=1}^nx_i \\
\hat{\beta_0} &= 0 - \hat{\beta_1}*\frac{1}{n}\sum_{i=1}^nx_i \\
\hat{\beta_0} &= - \hat{\beta_1}*(\bar{x}) \\
\hat{\beta_0} &= - \hat{\beta_1}\bar{x} \\
\end{align*}
$$

## A.2 

#### (a) Show that the ridge regression estimator is biased.

$$ 
\begin{align*}
\hat{\beta}_R &= (X^TX + \lambda I)^{-1}X^TY \\
&= (X^TX + \lambda I)^{-1}X^T * \Big[ (X)(X^TX + \lambda I)^{-1}X^T + (\epsilon) \Big]
\end{align*}
$$

That doesn't cancel nicely. In the case that $\lambda = 0$, then the estimate is the same as the estimator for ordinary least squares. However, when $\lambda > 0$, then the gram matrix will no longer cancel and there will be some bias in the estimator.

#### (b) Let $\boldsymbol{\widehat\beta}_{\lambda}$ denote the ridge regression estimator for a given $\lambda$, and let $\boldsymbol{\widehat\beta}$ denote the ordinary least squares estimator. It turns out that, if the original predictors are orthogonal and standardized, then:

$$\boldsymbol{\widehat\beta}_{\lambda}= \frac{1}{1 + \lambda}\boldsymbol{\widehat\beta},$$

and 

$$Var(\boldsymbol{\widehat\beta}_{\lambda}) = \bigg(\frac{1}{1+\lambda}\bigg)^2\sigma^2I_p,$$

#### where $I_p$ is the $p \times p$ identity matrix (You don't have to, but can you prove these two formulas??) Use these formulas to:

1. Justify the ridge regression estimator as a "shrinkage" method in two ways.
2. Describe what happens to the ridge regression estimator as $\lambda \to \infty$.

1. The first way that ridge regression is a shrinkage method is by introducing variance to reduce bias. By applying the ridge regression penalty, we are introducing bias to the estimator but reducing the estimator to zero, reducing variance.
2. We can see from the first formula that the ridge regression estimator will approach zero as $\lambda$ approaches infinity.

#### (c) Describe why we standardize our variables before performing ridge regression.

We can see in the equation for the ridge regression estimator that the penalty term is applied to all variables uniformly (the $+ \lambda I$ part). This means, if the variables weren't standardized, then the effect of the penalty term would unevenly affect the coefficients. When the variables are standardized, the penalty term applies to each one by the same amount

#### (d) Let $Y = \widehat\beta_0 + \widehat\beta_1X$ be the fitted model for unstandardized variables and $Z_2 = \widehat\theta Z_1$ be the model for the standardized model (predictors and response, i.e., $Z_2 = \frac{Y - \bar{Y}}{s_y}$, and $Z_1 = \frac{X - \bar{X}}{s_X}$). Construct a transformation to go from $\widehat\beta$ and $\widehat\theta$.

Let's start with the unstandardized model equation and work up from there.
$$
\begin{align*}
    Y &= \hat{\beta_0} + \hat{\beta_1}X \\
    Y - \bar{Y} &= \hat{\beta_0} + \hat{\beta_1}X - \bar{Y} \\
    Y - \bar{Y} &= (\bar{Y} - \hat{\beta_1}\bar{X}) + \hat{\beta_1}X - \bar{Y} \\
    Y - \bar{Y} &= - \hat{\beta_1}\bar{X} + \hat{\beta_1}X \\
    Y - \bar{Y} &= \hat{\beta_1}(X - \bar{X}) \\
    \dfrac{s_Y}{s_Y} (Y - \bar{Y}) &= \hat{\beta_1}(X - \bar{X}) \dfrac{s_X}{s_X}\\
    \dfrac{Y - \bar{Y}}{s_y} &= \dfrac{s_X}{s_Y} \hat{\beta_1} * \frac{X - \bar{X}}{s_X} \\
    Z_2 &= \dfrac{s_X}{s_Y} \hat{\beta_1} * Z_1
\end{align*}
$$

Therefor $\hat{\theta} = \dfrac{s_X}{s_Y} \hat{\beta_1}$.

## B. Computational Problems

### Problem B.1

#### (a) Using the divusa data, fit a regression of with divorce as the response and unemployed, femlab, marriage, birth, and military as predictors. Compute the variance inflation factors VIFs for each $\widehat\beta_j$, $j = 1,...,p$. Is there evidence that collinearity causes some predictors not to be significant? Explain your answer.



In [7]:
library(faraway) #vif() will calculate variance inflation factors...
data(divusa)

lmod = lm(divorce ~ . - year, data = divusa)
summary(lmod)
vif(lmod)


Call:
lm(formula = divorce ~ . - year, data = divusa)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.8611 -0.8916 -0.0496  0.8650  3.8300 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  2.48784    3.39378   0.733   0.4659    
unemployed  -0.11125    0.05592  -1.989   0.0505 .  
femlab       0.38365    0.03059  12.543  < 2e-16 ***
marriage     0.11867    0.02441   4.861 6.77e-06 ***
birth       -0.12996    0.01560  -8.333 4.03e-12 ***
military    -0.02673    0.01425  -1.876   0.0647 .  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.65 on 71 degrees of freedom
Multiple R-squared:  0.9208,	Adjusted R-squared:  0.9152 
F-statistic: 165.1 on 5 and 71 DF,  p-value: < 2.2e-16


unemployed     femlab   marriage      birth   military 
  2.252888   3.613276   2.864864   2.585485   1.249596

The variance inflation factors of each variable is relatively low, as none of the factors exceed 4, meaning there is not significant correlation between any of the variables. We can also see that most of the standard errors of the model are similar and not that large (not including the intercept), which leads to us thinking that they are significant. However, the p-values of the unemployed and military variables are slightly above the 95% level, so we could remove those variables and see how the model fits.

#### (b) Does the removal of insignificant predictors from the model reduce the collinearity? Investigate.

In [8]:
lmod2 = lm(divorce ~ femlab + marriage + birth, data = divusa)
summary(lmod2)
vif(lmod2)


Call:
lm(formula = divorce ~ femlab + marriage + birth, data = divusa)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.6923 -1.1934 -0.0534  1.2265  3.6701 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -1.54545    2.21247  -0.699    0.487    
femlab       0.41337    0.02275  18.174  < 2e-16 ***
marriage     0.12609    0.02199   5.735 2.07e-07 ***
birth       -0.11627    0.01412  -8.235 5.10e-12 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.695 on 73 degrees of freedom
Multiple R-squared:  0.9141,	Adjusted R-squared:  0.9106 
F-statistic: 258.9 on 3 and 73 DF,  p-value: < 2.2e-16


femlab marriage    birth 
1.893390 2.201891 2.008469

By taking out the unemployed and military predictors, we reduced the collinearity of femlab from 3.61 to 1.89, which is a large change. The marriage and birth predictors changed by about 0.6 each, marriage from 2.86 to 2.20 and birth from 2.59 to 2.00, which is a smaller change. The amount in collinearity in the model was reduced by the removal of those predictors.

### Problem B.2

The salmonella data was collected in a salmonella reverse mutagenicity essay. The predictor is the dose level of quinoline and the response is the numbers of revertant colonies of TA98 salmonella observed on each of three replicate plates. 

#### (a) Show that a Poisson GLM is inadequate and that some overdispersion must be allowed for. 

In [9]:
library(faraway)
data(salmonella)
head(salmonella)

glmod = glm(colonies ~ dose, data = salmonella, family = poisson)
summary(glmod)
drop1(glmod, test = "Chisq")

colonies dose
1 15        0  
2 21        0  
3 29        0  
4 16       10  
5 18       10  
6 21       10


Call:
glm(formula = colonies ~ dose, family = poisson, data = salmonella)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.6482  -1.8225  -0.2993   1.2917   5.1861  

Coefficients:
             Estimate Std. Error z value Pr(>|z|)    
(Intercept) 3.3219950  0.0540292  61.485   <2e-16 ***
dose        0.0001901  0.0001172   1.622    0.105    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 78.358  on 17  degrees of freedom
Residual deviance: 75.806  on 16  degrees of freedom
AIC: 172.34

Number of Fisher Scoring iterations: 4


Df Deviance AIC      LRT      Pr(>Chi) 
<none> NA 75.80589 172.3421       NA        NA
dose    1 78.35758 172.8938 2.551689 0.1101764

From the summary, we can see that the GLM is inadequate. First of all, the difference between the null deviance and residual deviance is small, meaning that fitting the model did not do much to address the variance in the data than just an intercept. From the drop1 function, we can also see that the dose predictor is not significant and increases the AIC of the model, meaning it can be removed.

#### (b) Estimate the overdispersion parameter and adjust the standard errors accordingly. 

In [10]:
dp = sum(resid(glmod, type = "pearson")^2) / glmod$df.res;
dp
summary(glmod, dispersion = dp)

[1] 5.087258


Call:
glm(formula = colonies ~ dose, family = poisson, data = salmonella)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.6482  -1.8225  -0.2993   1.2917   5.1861  

Coefficients:
             Estimate Std. Error z value Pr(>|z|)    
(Intercept) 3.3219950  0.1218626  27.260   <2e-16 ***
dose        0.0001901  0.0002644   0.719    0.472    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 5.087258)

    Null deviance: 78.358  on 17  degrees of freedom
Residual deviance: 75.806  on 16  degrees of freedom
AIC: 172.34

Number of Fisher Scoring iterations: 4


### Problem B.3


Consider the fat dataset, with percentage of body fat, siri, as the response and the other variables, except brozek and density, as potential predictors. 

#### Remove every tenth observation from the data for use as a test sample. Use the remaining data as a training sample building the following models:

1. Linear regression with all mentioned predictors.
2. Linear regression with variables selected using AIC.
3. Principal component regression.
4. Ridge regression.

#### Use the models you find to predict the response in the test sample. Make a report on the performance of the models. 

In [11]:
head(fat)
# Find the number of rows of the data
n = nrow(fat)
# Create a sequence of every 10 words up to the number of rows
s = seq(0, n, 10)
# Split into training and testing sets of every 10th row.
test = fat[s, ]
test = test[ , -c(1, 3)]
train = fat[-s, ]
train = train[ , -c(1, 3)]
# Fit the model on the training data
lmod = lm(siri ~ . , data = train)
# summary(lmod)

# Predict the response of the test set using the model
predMLR = predict(lmod, newdata = test)
# Calculate the mean squared error to see how well its doing
mseMLR = mean((test$siri - predMLR)^2)
mseMLR

brozek siri density age weight height adipos free  neck chest abdom hip  
1 12.6   12.3 1.0708  23  154.25 67.75  23.7   134.9 36.2  93.1  85.2  94.5
2  6.9    6.1 1.0853  22  173.25 72.25  23.4   161.3 38.5  93.6  83.0  98.7
3 24.6   25.3 1.0414  22  154.00 66.25  24.7   116.0 34.0  95.8  87.9  99.2
4 10.9   10.4 1.0751  26  184.75 72.25  24.9   164.7 37.4 101.8  86.4 101.2
5 27.8   28.7 1.0340  24  184.25 71.25  25.6   133.1 34.4  97.3 100.0 101.9
6 20.6   20.9 1.0502  24  210.25 74.75  26.5   167.0 39.0 104.5  94.4 107.8
  thigh knee ankle biceps forearm wrist
1 59.0  37.3 21.9  32.0   27.4    17.1 
2 58.7  37.3 23.4  30.5   28.9    18.2 
3 59.6  38.9 24.0  28.8   25.2    16.6 
4 60.1  37.3 22.8  32.4   29.4    18.2 
5 63.2  42.2 24.0  32.2   27.7    17.7 
6 66.0  42.0 25.6  35.7   30.6    18.8

[1] 1.280357

We can use the mean squared error (MSE) between the true test values and our predicted values to get an approximation for how well our model is fitting the data. For the base MLR model with all the predictors, we get an MSE of 1.280. This value doesn't mean much on its own but will be useful when comparing the fits of the other models.

In [12]:
library(leaps)
library(MASS)
# Create a model with predictors from the AIC model
lmodAIC = stepAIC(lmod)
summary(lmodAIC)

# Predict the response with the AIC model
predAIC = predict(lmodAIC, newdata = test)
# Compute the mean squared error to determine how well model is doing
mseAIC = mean((test$siri - predAIC)^2)
mseAIC

Start:  AIC=214.36
siri ~ age + weight + height + adipos + free + neck + chest + 
    abdom + hip + thigh + knee + ankle + biceps + forearm + wrist

          Df Sum of Sq    RSS    AIC
- hip      1       0.0  506.9 212.37
- neck     1       0.1  507.0 212.39
- age      1       1.0  507.9 212.81
- wrist    1       1.1  508.0 212.84
- knee     1       3.1  510.0 213.75
- height   1       3.6  510.4 213.94
<none>                  506.9 214.36
- biceps   1       5.3  512.2 214.73
- ankle    1       5.7  512.6 214.89
- chest    1      22.2  529.0 222.07
- forearm  1      23.8  530.7 222.77
- abdom    1      26.5  533.4 223.92
- thigh    1      30.8  537.7 225.75
- adipos   1      48.8  555.7 233.21
- weight   1     582.4 1089.3 386.01
- free     1    3456.8 3963.7 679.21

Step:  AIC=212.37
siri ~ age + weight + height + adipos + free + neck + chest + 
    abdom + thigh + knee + ankle + biceps + forearm + wrist

          Df Sum of Sq    RSS    AIC
- neck     1       0.1  507.0 210.40
- age


Call:
lm(formula = siri ~ weight + adipos + free + chest + abdom + 
    thigh + knee + ankle + biceps + forearm, data = train)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.7926 -0.6839  0.2371  0.8824  6.8655 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -7.24766    3.94171  -1.839 0.067331 .  
weight       0.36973    0.01994  18.546  < 2e-16 ***
adipos      -0.57022    0.09459  -6.028 7.09e-09 ***
free        -0.55965    0.01400 -39.978  < 2e-16 ***
chest        0.12099    0.03809   3.176 0.001709 ** 
abdom        0.15824    0.03908   4.049 7.18e-05 ***
thigh        0.16140    0.04551   3.546 0.000479 ***
knee         0.12767    0.08749   1.459 0.145947    
ankle        0.13817    0.07858   1.758 0.080106 .  
biceps       0.11222    0.06317   1.777 0.077055 .  
forearm      0.24281    0.06990   3.474 0.000620 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.544 on 216 degrees of freedom
Mu

[1] 1.258928

The AIC model selection method gives us a mean squared error of 1.258, which is somehow less than the full model. Which is wierd. Because the MLR models has more predictors, it should be a better fit than the AIC model. Ignoring that, the MSE tells us that the AIC model does fit the data well while being simpler than the MLR model.

In [13]:
# Fit a model using PCR
# Create PCs
s = prcomp(train[ , -1], scale = T)
# Create a new table of result and the PC values
df = data.frame(train$siri, s$x)
# Fit a model frmo the PCs to the result
lmodPCR = lm(train.siri ~ ., data = df)
# summary(prcomp(train[ , -c(1,2,3)], scale = T))
summary(prcomp(train[ , -1]), scale = T)

eigen = s$rotation
means = colMeans(train[, -1])
sds = apply(train[, -1], 2, sd)
testScaled = scale(test[ , -1], scale = sds, center = means)
testPC = testScaled %*% eigen
dfTest = data.frame(test$siri, testPC)
# Predict the response with the PCR model
predPCR = predict(lmodPCR, newdata = dfTest)
# Compute the mean squared error to determine how well the model is doing
msePCR = mean((test$siri - predPCR)^2)
msePCR

Warning message:
"In summary.prcomp(prcomp(train[, -1]), scale = T) :
 extra argument 'scale' will be disregarded"

Importance of components:
                           PC1     PC2     PC3     PC4     PC5     PC6     PC7
Standard deviation     37.0474 14.5403 9.51132 3.63395 3.46200 2.54427 2.19986
Proportion of Variance  0.7961  0.1226 0.05247 0.00766 0.00695 0.00375 0.00281
Cumulative Proportion   0.7961  0.9187 0.97115 0.97881 0.98576 0.98952 0.99233
                           PC8     PC9    PC10    PC11    PC12    PC13   PC14
Standard deviation     1.85387 1.57678 1.41170 1.32373 1.24936 1.03687 0.8285
Proportion of Variance 0.00199 0.00144 0.00116 0.00102 0.00091 0.00062 0.0004
Cumulative Proportion  0.99432 0.99576 0.99692 0.99793 0.99884 0.99946 0.9999
                          PC15
Standard deviation     0.49086
Proportion of Variance 0.00014
Cumulative Proportion  1.00000

[1] 1.280357

In [14]:
lmodPCR2 = lm(train.siri ~ PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7, data = df)
# Predict the response with the PCR model
predPCR2 = predict(lmodPCR2, newdata = dfTest)
# Compute the mean squared error to determine how well the model is doing
mean((test$siri - predPCR2)^2)

lmodPCR3 = lm(train.siri ~ PC1 + PC2, data = df)
# Predict the response with the PCR model
predPCR3 = predict(lmodPCR3, newdata = dfTest)
# Compute the mean squared error to determine how well the model is doing
mean((test$siri - predPCR3)^2)

[1] 7.127634

[1] 12.12824

The mean squared error of the principal component regression model, when all the principle components, was 1.280 which is the same as the MSE for the MLR model. This makes sense because the PC regression is using all the same data, only transforming it into priciple components rather than predictors, so when we use all the principal components, we can only get as good of a fit as the MLR model. When we look at the principle components, we see that over 99% of the variance is included by the first seven PCs, which then gives us an MSE of $7.128$. When we take the first two PCs, which account for about 92% of the total variance, the MSE increases to $12.128$.

In [15]:
# Fit a model using ridge regression
library(ridge)
lmodRidge = linearRidge(siri ~ ., data = train)
summary(lmodRidge)

# Predict the response with the ridge model
predRidge = predict(lmodRidge, newdata = test)
# Compute the mean squared error to determine how well the model predicted
mseRidge = mean((test$siri - predRidge)^2)
mseRidge


Call:
linearRidge(formula = siri ~ ., data = train)


Coefficients:
              Estimate Scaled estimate Std. Error (scaled) t value (scaled)
(Intercept)  -46.63148              NA                  NA               NA
age            0.01339         2.54266             2.49408            1.019
weight         0.14456        65.11437             4.91825           13.239
height         0.15443         8.71858             2.34155            3.723
adipos        -0.09079        -5.04482             4.77756            1.056
free          -0.43355      -119.40524             3.45520           34.558
neck           0.05299         1.94946             3.45355            0.564
chest          0.14785        19.05275             4.25804            4.475
abdom          0.28830        47.30761             4.60364           10.276
hip            0.13035        14.32576             4.61340            3.105
thigh          0.19360        15.48606             4.07616            3.799
knee           0.28

[1] 1.773162

From the ridge regression model, we get a mean squared error of $1.773$. This is the highest of the models, but the values are still very close. The added bias of the ridge regression did not significantly reduce the variance of the model.

### Problem B.4

Use the seatpos data with hipcenter as the response.

#### (a) Fit a model with all eight predictors. Comment on the effect of leg length on the response. 

In [16]:
library(faraway)
data(seatpos)

lmod = lm(hipcenter ~ ., data = seatpos)
summary(lmod)
drop1(lmod)
vif(lmod)


Call:
lm(formula = hipcenter ~ ., data = seatpos)

Residuals:
    Min      1Q  Median      3Q     Max 
-73.827 -22.833  -3.678  25.017  62.337 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)  
(Intercept) 436.43213  166.57162   2.620   0.0138 *
Age           0.77572    0.57033   1.360   0.1843  
Weight        0.02631    0.33097   0.080   0.9372  
HtShoes      -2.69241    9.75304  -0.276   0.7845  
Ht            0.60134   10.12987   0.059   0.9531  
Seated        0.53375    3.76189   0.142   0.8882  
Arm          -1.32807    3.90020  -0.341   0.7359  
Thigh        -1.14312    2.66002  -0.430   0.6706  
Leg          -6.43905    4.71386  -1.366   0.1824  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 37.72 on 29 degrees of freedom
Multiple R-squared:  0.6866,	Adjusted R-squared:  0.6001 
F-statistic:  7.94 on 8 and 29 DF,  p-value: 1.306e-05


Df Sum of Sq   RSS      AIC     
<none>  NA          NA 41261.78 283.6240
Age      1 2632.116916 43893.90 283.9739
Weight   1    8.993219 41270.78 281.6323
HtShoes  1  108.430657 41370.21 281.7238
Ht       1    5.014051 41266.80 281.6286
Seated   1   28.642794 41290.43 281.6504
Arm      1  164.974942 41426.76 281.7757
Thigh    1  262.761526 41524.55 281.8652
Leg      1 2654.848346 43916.63 283.9936

Age     Weight    HtShoes         Ht     Seated        Arm      Thigh 
  1.997931   3.647030 307.429378 333.137832   8.951054   4.496368   2.762886 
       Leg 
  6.694291

The estimate of Leg is -6.44, meaning that every incease in leg by 1 decreases hipcenter by 6.44. The p-value of this estimate is above 0.05, so we can say that it is does not signficantly affect the result, but all of the predictors are not-significant, and Leg has the lowest p-value, so that might be important. When we look at the variance inflation factors, we can see that there is some correlation between leg length and other variables, so we can take that into account in fitting the model.

#### (b) Compute a 95% prediction interval for the response at the mean value of the predictors. 

In [17]:
means = colMeans(seatpos)
means = data.frame(t(means))


pred = predict(lmod, newdata = means, interval = "prediction")
pred

fit       lwr     upr      
1 -164.8849 -243.04 -86.72972

The 95% prediction interval for the full model is [-243.04, -86.73].

#### (c) Use AIC to select a model. Now interpret the effect of leg length and compute the prediction interval. Compare the conclusions from the two models. 

In [18]:
library(MASS)
lmodAIC = stepAIC(lmod)
summary(lmodAIC)
vif(lmodAIC)

Start:  AIC=283.62
hipcenter ~ Age + Weight + HtShoes + Ht + Seated + Arm + Thigh + 
    Leg

          Df Sum of Sq   RSS    AIC
- Ht       1      5.01 41267 281.63
- Weight   1      8.99 41271 281.63
- Seated   1     28.64 41290 281.65
- HtShoes  1    108.43 41370 281.72
- Arm      1    164.97 41427 281.78
- Thigh    1    262.76 41525 281.87
<none>                 41262 283.62
- Age      1   2632.12 43894 283.97
- Leg      1   2654.85 43917 283.99

Step:  AIC=281.63
hipcenter ~ Age + Weight + HtShoes + Seated + Arm + Thigh + Leg

          Df Sum of Sq   RSS    AIC
- Weight   1     11.10 41278 279.64
- Seated   1     30.52 41297 279.66
- Arm      1    160.50 41427 279.78
- Thigh    1    269.08 41536 279.88
- HtShoes  1    971.84 42239 280.51
<none>                 41267 281.63
- Leg      1   2664.65 43931 282.01
- Age      1   2808.52 44075 282.13

Step:  AIC=279.64
hipcenter ~ Age + HtShoes + Seated + Arm + Thigh + Leg

          Df Sum of Sq   RSS    AIC
- Seated   1     35.10 4131


Call:
lm(formula = hipcenter ~ Age + HtShoes + Leg, data = seatpos)

Residuals:
    Min      1Q  Median      3Q     Max 
-79.269 -22.770  -4.342  21.853  60.907 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 456.2137   102.8078   4.438 9.09e-05 ***
Age           0.5998     0.3779   1.587   0.1217    
HtShoes      -2.3023     1.2452  -1.849   0.0732 .  
Leg          -6.8297     4.0693  -1.678   0.1024    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 35.13 on 34 degrees of freedom
Multiple R-squared:  0.6813,	Adjusted R-squared:  0.6531 
F-statistic: 24.22 on 3 and 34 DF,  p-value: 1.437e-08


Age  HtShoes      Leg 
1.011467 5.777752 5.751729

In [19]:
predict(lmodAIC, newdata = means, interval = "predict")

fit       lwr      upr      
1 -164.8849 -237.209 -92.56072

In [20]:
anova(lmodAIC, lmod, test = "F")

Res.Df RSS      Df Sum of Sq F          Pr(>F)   
1 34     41957.79 NA       NA          NA        NA
2 29     41261.78  5 696.0084  0.09783506 0.9917649

With the reduced model, we see that the estimated coefficient is -6.83, meaning that each increase in left length by 1 would decrease hipcenter by 6.83, which is greater greater than the full model. The p-value of the leg-estimate is also less in the reduced model than the full model, from 0.182 to 0.102. The p-value is still high for the value to be significicant, which is likely because of the high variance inflation factor of Leg, which is again less than in the full model. The prediction interval is [-237.21, -92.56] with a fit of -164.88. With an ANOVA test, we get a p-value of 0.992, which strongly supports that the reduced model is the same as the full model, so we should use the reduced model.

### Problem B.5

#### Use the teengamb data with gamble as the response and the other variables as predictors. Implement the following variable selection methods to determine the "best" model:

1. Backward Elimination
2. AIC
3. Adjusted $R^2$.

In [27]:
library(faraway)
data(teengamb)
# Backward selection
lmod = lm(gamble ~ ., data = teengamb)
summary(lmod)
# First step
cc = coef(summary(lmod))  # coefficient table
lnSigVariable = which.max(cc[,"Pr(>|t|)"]) # Get row with largest p-value
lnSigVariable
lmod2 = update(lmod, . ~ . - status)
summary(lmod2)


Call:
lm(formula = gamble ~ ., data = teengamb)

Residuals:
    Min      1Q  Median      3Q     Max 
-51.082 -11.320  -1.451   9.452  94.252 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)  22.55565   17.19680   1.312   0.1968    
sex         -22.11833    8.21111  -2.694   0.0101 *  
status        0.05223    0.28111   0.186   0.8535    
income        4.96198    1.02539   4.839 1.79e-05 ***
verbal       -2.95949    2.17215  -1.362   0.1803    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 22.69 on 42 degrees of freedom
Multiple R-squared:  0.5267,	Adjusted R-squared:  0.4816 
F-statistic: 11.69 on 4 and 42 DF,  p-value: 1.815e-06


status 
     3


Call:
lm(formula = gamble ~ sex + income + verbal, data = teengamb)

Residuals:
    Min      1Q  Median      3Q     Max 
-50.639 -11.765  -1.594   9.305  93.867 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  24.1390    14.7686   1.634   0.1095    
sex         -22.9602     6.7706  -3.391   0.0015 ** 
income        4.8981     0.9551   5.128 6.64e-06 ***
verbal       -2.7468     1.8253  -1.505   0.1397    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 22.43 on 43 degrees of freedom
Multiple R-squared:  0.5263,	Adjusted R-squared:  0.4933 
F-statistic: 15.93 on 3 and 43 DF,  p-value: 4.148e-07


In [23]:
# AIC model
lmodAIC = stepAIC(lmod)
summary(lmodAIC)

Start:  AIC=298.18
gamble ~ sex + status + income + verbal

         Df Sum of Sq   RSS    AIC
- status  1      17.8 21642 296.21
<none>                21624 298.18
- verbal  1     955.7 22580 298.21
- sex     1    3735.8 25360 303.67
- income  1   12056.2 33680 317.00

Step:  AIC=296.21
gamble ~ sex + income + verbal

         Df Sum of Sq   RSS    AIC
<none>                21642 296.21
- verbal  1    1139.8 22781 296.63
- sex     1    5787.9 27429 305.35
- income  1   13236.1 34878 316.64



Call:
lm(formula = gamble ~ sex + income + verbal, data = teengamb)

Residuals:
    Min      1Q  Median      3Q     Max 
-50.639 -11.765  -1.594   9.305  93.867 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  24.1390    14.7686   1.634   0.1095    
sex         -22.9602     6.7706  -3.391   0.0015 ** 
income        4.8981     0.9551   5.128 6.64e-06 ***
verbal       -2.7468     1.8253  -1.505   0.1397    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 22.43 on 43 degrees of freedom
Multiple R-squared:  0.5263,	Adjusted R-squared:  0.4933 
F-statistic: 15.93 on 3 and 43 DF,  p-value: 4.148e-07


In [32]:
# Find model with highest adjusted r squared
summary(lm(gamble ~ ., data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - sex, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - income, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - status, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - verbal, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - sex - income, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - sex - status, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - sex - verbal, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - income - status, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - income - verbal, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - status - verbal, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - sex - income - status, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - sex - income - verbal, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - sex - status - verbal, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - income - status - verbal, data = teengamb))$adj.r.squared
summary(lm(gamble ~ . - sex - income - status - verbal, data = teengamb))$adj.r.squared

[1] 0.4816495

[1] 0.4062348

[1] 0.211421

[1] 0.4932879

[1] 0.4713267

[1] 0.01162814

[1] 0.372367

[1] 0.3755844

[1] 0.2019398

[1] 0.211081

[1] 0.478724

[1] 0.02727861

[1] -0.01962347

[1] 0.373357

[1] 0.1477786

[1] 0

All three methods agree that the having the predictors sex, income and verbal is the best fit for the model with the least number of predictors. The backward selection has all p-values of predictors under $\alpha = 0.15,$ the AIC function stopped at the three predictors, and the three predictors had the highest adjusted $R^2$ value of 0.493.